cpp/oneapi/dal/algo/knn/backend/cpu/train_kernel_kd_tree.cpp

```diff
@@ -0,0 +1,102 @@
+* Copyright 2020 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "data_management/data/numeric_table.h"
+#include <daal/src/algorithms/k_nearest_neighbors/kdtree_knn_classification_train_kernel.h>
+#include "src/algorithms/k_nearest_neighbors/kdtree_knn_classification_model_impl.h"
+#include "algorithms/engines/mcg59/mcg59.h"
+
+#include "oneapi/dal/backend/interop/common.hpp"
+#include "oneapi/dal/backend/interop/table_conversion.hpp"
+#include "oneapi/dal/algo/knn/backend/cpu/train_kernel.hpp"
+#include "oneapi/dal/algo/knn/detail/model_impl.hpp"
+#include "oneapi/dal/algo/knn/backend/model_interop.hpp"
+
+namespace oneapi::dal::knn::backend {
+
+using std::int64_t;
+using dal::backend::context_cpu;
+using namespace daal::data_management;
+
+namespace daal_knn = daal::algorithms::kdtree_knn_classification;
+namespace interop  = dal::backend::interop;
+
+template <typename Float, daal::CpuType Cpu>
+using daal_knn_kd_tree_kernel_t = daal_knn::training::internal::KNNClassificationTrainBatchKernel<Float, 
+                    daal_knn::training::defaultDense, Cpu>;
+
+template <typename Float>
+static train_result call_daal_kernel(const context_cpu& ctx,
+                                     const descriptor_base& desc,
+                                     const table& data,
+                                     const table& labels) {
+    const int64_t row_count = data.get_row_count();
+    const int64_t column_count = data.get_column_count();
+
+    auto arr_data = row_accessor<const Float>{ data }.pull();
+    auto arr_labels = row_accessor<const Float>{ labels }.pull();
+    // TODO: read-only access performed with deep copy of data since daal numeric tables are mutable.
+    // Need to create special immutable homogen table on daal interop side
+
+    // TODO: data is table, not a homogen_table. Think better about accessor - is it enough to have just a row_accessor?
+    const auto daal_data = interop::convert_to_daal_homogen_table(arr_data, row_count, column_count);
+    const auto daal_labels = interop::convert_to_daal_homogen_table(arr_labels, row_count, 1);
+
+
+    daal_knn::Parameter daal_parameter(desc.get_class_count(),
+        desc.get_neighbor_count(), desc.get_seed(), desc.get_data_use_in_model() ? daal_knn::doUse : daal_knn::doNotUse);
+
+    const daal_knn::ModelPtr model_ptr = 
```

Comment:
I'd recommend to use `auto` if the type of right hand side expression is clear from context for the reader 
```suggestion
    const auto model_ptr = 
```

Ignore it, if you don't feel in the same way 😄